src/elements/frame-router.ts

```diff
@@ -126,6 +128,13 @@ class FrameRouterElement extends HTMLElement {
   }
 
   private _handleLifecycleMessage(message: LabeledStarted) {
+    this._frameManager.sendToClient({
```

Comment:
This sends a message to the client with the id that the host thinks it is.
I used a confirmationId since we could in theory have two clients going through initialization, so its just a way to make sure that it confirms that it got the right client.

Although, this still doesn't solve the issue fully since the host doesn't know which temp id is which...  Maybe this situation isn't worth programming around?

Alternative to this is that we could have the temp id be the id passed from the client be permanent.  Then the FrameRouter can reconcile the random id to the configured id.

Reply:
Any reason to not include this in env init message?

Also, it occurs to me now that maybe we just send the id used to register the clients?  So that the host app can make the decisions without an opaque key?  I guess there's a chance where the host doesn't want the client to have it's id, but that seems unlikely and not worth the trouble of mapping between the key spaces.

Reply:
I saw this, thought I initially understood it, commented on it, and now I'm starting to question it all 😱

I was thinking of lots of edge cases.  The one you pointed out that an init procedure can be interrupted at any time by a user navigating to another client that results in a page load.  Also, there's the case where the frame is using hash routing internally, but not playing nice with our routing and navigating between the pages.

As such, I'm wondering if it even makes sense to try and exchange IDs or track them.  Maybe, the best approach is to just let the message come in and be decorate userspace events (events handled by the host embedding app) with the current client id and url.  We could also dump the current client id if the frame has been navigated and doesn't match the expected origin.

In this way, user land can still know the client id of the (toast, nav, etc.) requestor, and we dont have to worry about state and race conditions.

Suspect we'll have to chat a bit more about this, but just wanted to get some thoughts out.

Reply:
I'm not sure the client needs to be in the loop about what it's id is. We also don't want the host to have to trust the client apps to not send incorrect identifiers.  The `FrameManager` class already knows what domain it expects the client to be on and doesn't allow any messages that aren't from that domain. We should be able to track the expected client id on the host side, and add it there.

If we want additional protections against accidentally mixing up same-domain clients, we could even validate the current location of the client window against the expected client path when tagging the message with the client id.  Technically it's possible that the client frame navigated after a valid message was sent and we could drop messages that should be valid, but the client frames aren't supposed to navigate on their own anyway.  It shouldn't be an issue for a client that is behaving correctly.

Reply:
I dropped the client_id passing to the client.  Now frame-router will add the "current client id" to the messages.